src/utils/format_stackdriver/format_stackdriver.c

```diff
@@ -134,16 +134,24 @@ static int format_typed_value(yajl_gen gen, metric_t const *m,
 
   switch (m->family->type) {
   case METRIC_TYPE_GAUGE: {
-    int status = json_string(gen, "doubleValue");
-    if (status != 0)
+    int status = json_string(gen, "doubleValue") ||
+                 (int)yajl_gen_double(gen, m->value.gauge);
+    if (status != 0) {
```

Comment:
Although system `yajl_gen.h` header seems to declare `yajl_gen_status_ok` also as zero, IMHO this change does _not_ improve the code.

Additionally, this function seems to leak `yajl_gen_map` on errors => set error value and break from switch, only then return error value?

Reply:
This is how the rest of the file avoids to check each call for error individually.

What do you think about the `ERR_COMBINE()` macro in `src/utils/format_influxdb/format_influxdb.c`? That would be a possible alternative.

Reply:
`yajl_gen_map_open` basically pushes a `{` into the `yajl_gen` generator, which is freed later. So there is not leak. But …

… the early return leaves the generator in an unknown and likely broken state, such as an opened map that we will never close. This happens not just here, but all throughout the file. I just pushed a change that will reset the generator state in case of an error so that we're back in a well defined state.

Reply:
Moved this into  #4268, because we need to cherry-pick this into 5.12 / main also.

Reply:
> This is how the rest of the file avoids to check each call for error individually.

Why the variable is `int`?  `||` operator reduces result to boolean value, i.e. looses the actual integer return value.

> What do you think about the ERR_COMBINE() macro in src/utils/format_influxdb/format_influxdb.c? That would be a possible alternative.

As that macro is only given ints, it could be function instead (`static inline` one in a header), e.g:
```
/* return non-zero one of the given two return values */
static int return_nonzero(int ret1, int ret2);
/* or */
static int nonzero_error(int ret1, int ret2);
```

Reply:
I removed the `int` cast.

Using a function does result in the same behavior: With the macro, the second argument is evaluated lazily, i.e. the function is only called if `!err`. With the `nonzero_error` function above, both arguments need to be evaluated before the function is called.

Reply:
> I removed the int cast.

My point was that result of `int || int` = `bool`, not `int`:
```
$ cat test.c 
#include <stdio.h>
int main() {
	int a = 2, b = 3;
	printf("%d || %d = %d\n", a, b, a||b);
	return 0;
}
$ gcc -O test.c -o test && ./test
2 || 3 = 1
```

> With the macro, the second argument is evaluated lazily, i.e. the function is only called if !err. With the nonzero_error function above, both arguments need to be evaluated before the function is called.

Macro is OK.  [1]

----

[1] I thought that function to be small enough that GCC could inline it, but but that does not seem to be the case:
```
$ cat test.c 
#include <stdio.h>
int print1(char *s) { return printf("1: %s\n", s); }
int print2(char *s) { return printf("2: %s\n", s); }
int print3(char *s) { return printf("3: %s\n", s); }
static inline int nonzero_error(int ret1, int ret2) { return ret1 || ret2; }
int main() {
	return nonzero_error(print1("always") || print2("direct-OR"), print3("indirect-OR"));
}
$ gcc -O3 -Wall test.c -o test && ./test
3: indirect-OR
1: always
```
Quick googling did not answer it, but maybe C evaluation rules require all arguments to be evaluated _even_ when compiler has full visibility to the called function?

Reply:
> Quick googling did not answer it, but maybe C evaluation rules require all arguments to be evaluated _even_ when compiler has full visibility to the called function?

Correct.